### PR TITLE
docs: Make 'Statement of Need' explicit section in JOSS paper

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

As requested by the JOSS Editor in Chief for the review https://github.com/openjournals/joss-reviews/issues/2823, add a section for "Statement of Need". Additionally, move the information on the archival of the `pyhf` source code on Zenodo and the version used in the review out of "Statement of Need" and (back into) the end of "Summary" as this information doesn't belong with motivational text.

For posterity, related PRs to the JOSS paper revision are:

- PR #1089 
- PR #1280 

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add 'Statement of Need' section heading to JOSS paper
   - c.f. https://github.com/openjournals/joss-reviews/issues/2823
* Move information on version and Zenodo archive back to 'Summary' section
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
@kratsg @lukasheinrich @cranmer If you can review these changes to the paper:

1. Go to https://whedon.theoj.org/
2. Paper repository address: `https://github.com/scikit-hep/pyhf/`
3. Custom branch: `docs/make-statement-of-need-a-section`
4. Compile

I'll then tag the Editor and Editor in Chief for the review to get their approval that these changes will be acceptable for publication. We can then merge, update the JOSS review Issue, and request review of the paper again.

For a reference "Statement of Need" section c.f. the [JOSS Example paper and bibliography](https://joss.readthedocs.io/en/latest/submitting.html#example-paper-and-bibliography).

**Comment by matthewfeickert** [comment]
I put this into Draft mode to help protect it from getting merged until after we get Editor okays.

File: docs/JOSS/paper.md [hunk 1]
```diff
@@ -40,13 +40,15 @@ The relationship between them is often formalised in a statistical model $f(\mat
 Given observed data, the likelihood $\mathcal{L}(\mathbf{\phi})$ then serves as the basis for inference on the parameters $\mathbf{\phi}$.
 For measurements based on binned data (histograms), the `HistFactory` family of statistical models [@Cranmer:1456844] has been widely used in both Standard Model measurements [@HIGG-2013-02] as well as searches for new physics [@ATLAS-CONF-2018-041].
 `pyhf` is a pure-Python implementation of the `HistFactory` model specification and implements a declarative, plain-text format for describing `HistFactory`-based likelihoods that is targeted for reinterpretation and long-term preservation in analysis data repositories such as HEPData [@Maguire:2017ypu].
+The source code for `pyhf` has been archived on Zenodo with the linked DOI: [@pyhf_zenodo].
+At the time of writing this paper the most recent release of `pyhf` is [`v0.5.4`](https://doi.org/10.5281/zenodo.4318533).
```
Comment (kratsg) [inline docs/JOSS/paper.md:44]:
```suggestion
At the time of writing this paper, the most recent release of `pyhf` is [`v0.5.4`](https://doi.org/10.5281/zenodo.4318533).
```

**Review by kratsg — CHANGES_REQUESTED** [review]

**Review by kratsg — APPROVED** [review]

File: docs/JOSS/paper.md [hunk 2]
```diff
@@ -40,13 +40,15 @@ The relationship between them is often formalised in a statistical model $f(\mat
...
 For measurements based on binned data (histograms), the `HistFactory` family of statistical models [@Cranmer:1456844] has been widely used in both Standard Model measurements [@HIGG-2013-02] as well as searches for new physics [@ATLAS-CONF-2018-041].
 `pyhf` is a pure-Python implementation of the `HistFactory` model specification and implements a declarative, plain-text format for describing `HistFactory`-based likelihoods that is targeted for reinterpretation and long-term preservation in analysis data repositories such as HEPData [@Maguire:2017ypu].
+The source code for `pyhf` has been archived on Zenodo with the linked DOI: [@pyhf_zenodo].
+At the time of writing this paper, the most recent release of `pyhf` is [`v0.5.4`](https://doi.org/10.5281/zenodo.4318533).
+
+# Statement of Need
 
 Through adoption of open source "tensor" computational Python libraries, `pyhf` decreases the abstractions between a physicist performing an analysis and the statistical modeling without sacrificing computational speed.
 By taking advantage of tensor calculations, `pyhf` outperforms the traditional `C++` implementation of `HistFactory` on data from real LHC analyses.
```
[the remark this answers is not in the document]
Comment (cranmer) [reply docs/JOSS/paper.md:49]: well, I wrote this a while ago, but didn't click "submit review". Paper is published! So ignore.

[the remark this answers is not in the document]
Comment (kratsg) [reply docs/JOSS/paper.md:49]: Yeah, I think we got stuck between vague requirements and having the JOSS take a while in this case. It is a fair balance to just repurpose existing text -- but your point is noted. I doubt we can make that phrasing change.

[the remark this answers is not in the document]
Comment (matthewfeickert) [reply docs/JOSS/paper.md:49]: Sorry this didn't get posted in time, but thanks for still adding this comment though! Still good to have for the future (as more JOSS papers will come from IRIS-HEP and Scikit-HEP).

**Pull request merged**